src/rascal/representations/calculator_spherical_expansion.hh

```diff
@@ -1716,35 +1729,34 @@ namespace rascal {
                                                                      neigh);
         double f_c{cutoff_function->f_c(dist)};
         auto coefficients_center_by_type{coefficients_center[neigh_type]};
-
-        // compute the coefficients
         size_t l_block_idx{0};
-        for (size_t angular_l{0}; angular_l < this->max_angular + 1;
-             ++angular_l) {
-          size_t l_block_size{2 * angular_l + 1};
-          c_ij_nlm.block(0, l_block_idx, max_radial, l_block_size) =
-              neighbour_contribution.col(angular_l) *
-              harmonics.segment(l_block_idx, l_block_size);
-          l_block_idx += l_block_size;
-        }
-        c_ij_nlm *= f_c;
-        coefficients_center_by_type += c_ij_nlm;
-
-        // half list branch for c^{ji} terms using
-        // c^{ij}_{nlm} = (-1)^l c^{ji}_{nlm}.
-        if (IsHalfNL) {
-          if (not manager->is_center_atom(atom_j)) {
-            std::stringstream err_str{};
-            err_str << "Half neighbor list should only be used when all the "
-                    << "atoms inside the unit cell are centers, i.e. "
-                    << "center_atoms_mask should not mask atoms.";
-            throw std::runtime_error(err_str.str());
+        // compute the coefficients
+        // complications here because the infrastructure for swaping ij is not
+        // here yet.
+        // TODO(felix) make swaping ij possible and implement the
+        // proper half for gradients so the whole computation can be skipped.
```

Comment:
So is half neighbourlist for the gradients turned off now? I remember discussing such a swap-ij function, but I thought it'd be part of this PR. Did you not end up needing it?

In any case, let's turn this into an issue so as not to forget about it.

Reply:
Btw, if this is half-neighbourlist code, where did the `IsHalfNL` test go? Is it now in `are_any_centers_masked`?

Reply:
For spherical invarients gradients we don't want to have a halfNL since \grad_i p^j =/= \grad_j p^i. So I removed the IsHalfNL test and made a static assert about it not being compatible in spherical invarients. Then we don't have HalfNL but we still want to do half of the work in spherical expansion, the problem is that for gradients we need a swap ij functionality to make it work. This is not part of this PR but I am currently working on such implementation in `feat/swap_pair`.